pkg/ddl/executor.go

```diff
@@ -888,32 +888,83 @@ func GetDefaultCollation(cs string, defaultUTF8MB4Collation string) (string, err
 // ResolveCharsetCollation will resolve the charset and collate by the order of parameters:
 // * If any given ast.CharsetOpt is not empty, the resolved charset and collate will be returned.
 // * If all ast.CharsetOpts are empty, the default charset and collate will be returned.
+// * This is used for:
+//   - CREATE TABLE
+//   - CREATE SCHEMA (aka CREATE DATABASE)
+//   - ALTER TABLE ... ADD COLUMN
+//   - ALTER TABLE ... MODIFY COLUMN
 func ResolveCharsetCollation(charsetOpts []ast.CharsetOpt, utf8MB4DefaultColl string) (chs string, coll string, err error) {
-	for _, v := range charsetOpts {
-		if v.Col != "" {
-			collation, err := collate.GetCollationByName(v.Col)
+	// for CREATE TABLE:
+	//   objectChs: Table character set, if specified
+	//   objectColl: Table collation, if specified
+	//   tempChs: Schema character set
+	//   tempColl: Schema collation
+	// for CREATE SCHEMA:
+	//   objectChs: schema character set, if specified
+	//   objectColl: schema collation, if specified
+	//   tempChs, tempColl: empty
+	var templChs, templColl string
+	objectChs := charsetOpts[0].Chs
+	objectColl := charsetOpts[0].Col
+	if len(charsetOpts) > 1 {
+		templChs = charsetOpts[1].Chs
+		templColl = charsetOpts[1].Col
+	}
```

Comment:
in `ProcessColumnCharsetAndCollation`, there might be 3 items in the slice

Reply:
Ok, so that's column, table, schema.

I'll make sure to add tests for that, and either make `ResolveCharsetCollation` handle this  or have `ProcessColumnCharsetAndCollation` use a slice of 2 items.

Reply:
I haven't found any situation yet in which the third item in `ProcessColumnCharsetAndCollation` is useful.